desc/compute/utils.py

```diff
@@ -1347,3 +1360,776 @@ def body(i, mins):
     # The above implementation was benchmarked to be more efficient than
     # alternatives without explicit loops in GitHub pull request #501.
     return grid.expand(mins, surface_label)
+
+
+def polyint(c):
+    """Coefficients for the primitives of the given set of polynomials.
+
+    Parameters
+    ----------
+    c : ndarray
+        First axis should store coefficients of a polynomial.
+        For a polynomial given by ∑ᵢⁿ cᵢ xⁱ, where n is ``c.shape[0] - 1``,
+        coefficient cᵢ should be stored at ``c[n - i]``.
+
+    Returns
+    -------
+    poly : ndarray
+        Coefficients of polynomial primitive, ignoring the arbitrary constant.
+        That is, ``poly[i]`` stores the coefficient of the monomial xⁿ⁻ⁱ⁺¹,
+        where n is ``c.shape[0] - 1``.
+
+    """
+    poly = (c.T / jnp.arange(c.shape[0], 0, -1)).T
+    return poly
+
+
+def polyder(c):
+    """Coefficients for the derivatives of the given set of polynomials.
+
+    Parameters
+    ----------
+    c : ndarray
+        First axis should store coefficients of a polynomial.
+        For a polynomial given by ∑ᵢⁿ cᵢ xⁱ, where n is ``c.shape[0] - 1``,
+        coefficient cᵢ should be stored at ``c[n - i]``.
+
+    Returns
+    -------
+    poly : ndarray
+        Coefficients of polynomial derivative, ignoring the arbitrary constant.
+        That is, ``poly[i]`` stores the coefficient of the monomial xⁿ⁻ⁱ⁻¹,
+        where n is ``c.shape[0] - 1``.
+
+    """
+    poly = (c[:-1].T * jnp.arange(c.shape[0] - 1, 0, -1)).T
+    return poly
+
+
+def polyval(x, c):
```

Comment:
couldn't this be done with some version of `jnp.vectorize(jnp.polyval)`? Same for `polyder` and `polyint`

Reply:
`jnp.polyint` pads the array with a zero, which is less convenient. vectorizing `jnp.polyder` should work, but jnp.vectorize is a loop if `jnp` defaults to numpy. I don't think `jnp.polyval` replicates this function. For example, `jnp.polyval` already has some vectorization to evaluate a single polynomial at a given array of points, so we would have to undo that vectorization and then call `jnp.vectorize` in some way to reimplement this. The `einsum` should be more efficient anyway.